src/IRM/IRMAdaptiveCurve.sol

```diff
@@ -0,0 +1,208 @@
+    /// @param _CURVE_STEEPNESS The steepness of the interest rate line.
+    /// @param _ADJUSTMENT_SPEED The speed at which the rate at target utilization is adjusted up or down.
+    constructor(
+        int256 _TARGET_UTILIZATION,
+        int256 _INITIAL_RATE_AT_TARGET,
+        int256 _MIN_RATE_AT_TARGET,
+        int256 _MAX_RATE_AT_TARGET,
+        int256 _CURVE_STEEPNESS,
+        int256 _ADJUSTMENT_SPEED
+    ) {
+        // Validate parameters.
+        if (_TARGET_UTILIZATION < 0 || _TARGET_UTILIZATION > 1e18) {
+            revert InvalidParams();
+        }
+        if (_INITIAL_RATE_AT_TARGET < _MIN_RATE_AT_TARGET || _INITIAL_RATE_AT_TARGET > _MAX_RATE_AT_TARGET) {
+            revert InvalidParams();
+        }
+        if (_MIN_RATE_AT_TARGET < 0.001e18 / YEAR || _MIN_RATE_AT_TARGET > 10e18 / YEAR) {
+            revert InvalidParams();
+        }
+        if (_MAX_RATE_AT_TARGET < 0.001e18 / YEAR || _MAX_RATE_AT_TARGET > 10e18 / YEAR) {
+            revert InvalidParams();
+        }
+        if (_MIN_RATE_AT_TARGET > _MAX_RATE_AT_TARGET) {
+            revert InvalidParams();
+        }
+        if (_CURVE_STEEPNESS < 1.01e18 || _CURVE_STEEPNESS > 100e18) {
+            revert InvalidParams();
+        }
+        if (_ADJUSTMENT_SPEED < 2e18 / YEAR || _ADJUSTMENT_SPEED > 1000e18 / YEAR) {
+            revert InvalidParams();
+        }
+
+        TARGET_UTILIZATION = _TARGET_UTILIZATION;
+        INITIAL_RATE_AT_TARGET = _INITIAL_RATE_AT_TARGET;
+        MIN_RATE_AT_TARGET = _MIN_RATE_AT_TARGET;
+        MAX_RATE_AT_TARGET = _MAX_RATE_AT_TARGET;
+        CURVE_STEEPNESS = _CURVE_STEEPNESS;
+        ADJUSTMENT_SPEED = _ADJUSTMENT_SPEED;
+    }
+
+    /// @inheritdoc IIRM
+    function computeInterestRate(address vault, uint256 cash, uint256 borrows) external returns (uint256) {
+        if (msg.sender != vault) revert E_IRMUpdateUnauthorized();
+        (uint256 avgRate, uint256 endRateAtTarget) = computeInterestRateInternal(vault, cash, borrows);
+        irState[vault] = IRState(uint208(endRateAtTarget), uint48(block.timestamp));
+        return avgRate * 1e9; // Extend rate to RAY/sec for EVK.
+    }
+
+    /// @inheritdoc IIRM
+    function computeInterestRateView(address vault, uint256 cash, uint256 borrows) external view returns (uint256) {
+        (uint256 avgRate,) = computeInterestRateInternal(vault, cash, borrows);
+        return avgRate * 1e9; // Extend rate to RAY/sec for EVK.
+    }
+
+    /// @notice Perform computation of the new rate at target without mutating state.
+    /// @param vault Address of the vault to compute the new interest rate for.
+    /// @param cash Amount of assets held directly by the vault.
+    /// @param borrows Amount of assets lent out to borrowers by the vault.
+    /// @return The new rate at target utilization in RAY units.
+    function computeRateAtTargetView(address vault, uint256 cash, uint256 borrows) external view returns (uint256) {
```

Comment:
I think I understand the averaging now, but I still don't understand the purpose of this accessor. What would we use it for? Morpho does not expose this value AFAICT.

Reply:
Ahhh well I see their `rateAtTarget` is actually public. But still, that is a different value than what this accessor computes (it's potentially out-of-date).

Reply:
OK sorry, I think I'm just being dumb. You'll need this value to actually plot the current interest rate graph. Makes sense now.